### PR TITLE
Leveldb

### PR DESCRIPTION
**Pull request by almogdepaz**
add level db to our mesh for persistence

### CODE REVIEW

File: database/db.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	db *leveldb.DB
+	wo *opt.WriteOptions
+	ro *opt.ReadOptions
+}
+
+func (db LevelDB) Close() {
+	db.db.Close()
+}
+
+func (db LevelDB) Put(key, value []byte) error {
```
Comment (antonlerner) [inline database/db.go:25]: in my opnion receiver should be of pointer type

File: mesh/mesh.go [hunk 2]
```diff
@@ -1,116 +1,129 @@
...
-func (ll *LayersDB) SetLatestKnownLayer(idx uint32) {
-	//todo thread safety
-	ll.latestKnownLayer = idx
-}
+func (m *mesh) AddLayer(layer *Layer) error {
+	m.lMutex.Lock()
+	defer m.lMutex.Unlock()
+	count := LayerID(m.LatestIrreversible())
+	if count > layer.Index() {
+		log.Debug("can't add layer ", layer.Index(), "(already exists)")
```
Comment (antonlerner) [inline mesh/mesh.go:74]: what if layers are not ordered?

File: mesh/mesh.go [hunk 3]
```diff
@@ -1,116 +1,129 @@
...
+	return m.latestKnownLayer
 }
 
-func (ll *LayersDB) LatestKnownLayer() uint32 {
-	//todo thread safety
-	return ll.latestKnownLayer
+func (m *mesh) SetLatestKnownLayer(idx uint32) {
+	defer m.lkMutex.Unlock()
+	m.lkMutex.Lock()
+	if idx > m.latestKnownLayer {
```
Comment (antonlerner) [inline mesh/mesh.go:63]: using atomic here will simplify code as well

File: mesh/mesh.go [hunk 4]
```diff
@@ -1,116 +1,129 @@
...
+}
 
+func (m *mesh) AddBlock(block *Block) error {
+	log.Debug("add block ", block.ID())
+	if err := m.mDB.addBlock(block); err != nil {
+		log.Debug("failed to add block ", block.ID(), " ", err)
+		return err
+	}
+	m.SetLatestKnownLayer(uint32(block.Layer()))
+	m.tortoise.HandleLateBlock(block) //todo should be thread safe?
```
Comment (antonlerner) [inline mesh/mesh.go:107]: yes, but lets discuss this

File: mesh/mesh.go [hunk 5]
```diff
@@ -1,116 +1,129 @@
...
+	return m.mDB.getBlock(id)
+}
+
+func (m *mesh) GetContextualValidity(id BlockID) (bool, error) {
+	return m.mDB.getContextualValidity(id)
+}
+
+func (m *mesh) Close() {
+	log.Debug("closing mDB")
+	m.mDB.Close()
```
Comment (antonlerner) [inline mesh/mesh.go:122]: should we keep a closed flag? otherwise what are the consequences of closing?

File: mesh/meshdb.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+	for _, b := range layer.blocks {
+		ids[b.Id] = true
+	}
+
+	//add blocks to mDB
+	for _, b := range layer.blocks {
+		bytes, err := blockAsBytes(*b)
+		if err != nil {
+			log.Error("problem serializing block ", b.ID(), err)
+			delete(ids, b.ID()) //remove failed block from layer
```
Comment (antonlerner) [inline mesh/meshdb.go]: This error is critical and should not be supressed

File: mesh/meshdb.go [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+			continue
+		}
+
+		err = m.blocks.Put(b.ID().ToBytes(), bytes)
+		if err != nil {
+			log.Error("could not add block ", b.ID(), " to database ", err)
```
Comment (antonlerner) [inline mesh/meshdb.go]: same here

File: mesh/meshdb.go [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+	}
+
+	return m.updateLayerIds(err, block)
+}
+
+func (m *meshDB) getLayerLock(index LayerID) sync.Mutex {
+	layerLock, found := m.layerLocks[index]
+	if !found {
+		layerLock = sync.Mutex{}
+		m.layerLocks[index] = layerLock
```
Comment (antonlerner) [inline mesh/meshdb.go]: leak - there is no mechanism for removing unused locks

File: sync/sync.go [hunk 9]
```diff
@@ -198,7 +202,7 @@ func (s *Syncer) getLayerBlockIDs(index uint32) chan uint32 {
 		m[string(hash)] = p
 	}
 
-	idSet := make(map[uint32]bool, 300) //todo move this to config
+	idSet := make(map[uint64]bool, 300) //todo move this to config
```
Comment (antonlerner) [inline sync/sync.go]: please move to config or at least name it

File: sync/sync.go [hunk 10]
```diff
@@ -213,24 +217,28 @@ func (s *Syncer) getLayerBlockIDs(index uint32) chan uint32 {
 		}
 	}
 
-	res := make(chan uint32, len(idSet))
+	res := make(chan mesh.BlockID, len(idSet))
```
Comment (antonlerner) [inline sync/sync.go]: typedef chan if using it as a queue

**Review by antonlerner — APPROVED** [review]

File: database/db.go (hunk 1 above)
Comment (almogdepaz) [reply database/db.go:25]: LevelDB holds only pointers

File: mesh/meshdb.go (hunk 6 above)
Comment (almogdepaz) [reply mesh/meshdb.go]: i don't think we should stop adding the whole layer because of a failed block

File: sync/sync.go (hunk 9 above)
Comment (almogdepaz) [reply sync/sync.go]: unnamed and not labeled because this goes to protobuff serialisation right now and protobuff dosent handle that very well

**Pull request merged**